tests/fixed_update.rs

```diff
@@ -0,0 +1,293 @@
+        0
+    );
+    assert_eq!(
+        app.world
+            .get_resource::<UpdateCounter>()
+            .unwrap()
+            .just_pressed,
+        1
+    );
+
+    // the FixedUpdate schedule should run once and the action still be counted as `just_pressed`
+    app.update();
+    assert_eq!(
+        app.world.get_resource::<FixedUpdateCounter>().unwrap().run,
+        1
+    );
+    assert_eq!(
+        app.world
+            .get_resource::<FixedUpdateCounter>()
+            .unwrap()
+            .just_pressed,
+        1
+    );
+    assert_eq!(
+        app.world
+            .get_resource::<UpdateCounter>()
+            .unwrap()
+            .just_pressed,
+        1
+    );
+
+    // the FixedUpdate schedule should run once, the button should now be `pressed`
+    app.update();
+    assert_eq!(
+        app.world.get_resource::<FixedUpdateCounter>().unwrap().run,
+        2
+    );
+    assert_eq!(
+        app.world
+            .get_resource::<FixedUpdateCounter>()
+            .unwrap()
+            .just_pressed,
+        1
+    );
+    assert_eq!(
+        app.world
+            .get_resource::<UpdateCounter>()
+            .unwrap()
+            .just_pressed,
+        1
+    );
+
+    // make sure that the timings didn't get updated twice (once in Update and once in FixedUpdate)
+    // (the `tick` function has been called twice, but it uses `Time<Real>` to update the time,
+    // which is only updated in `PreUpdate`, which is what we want)
+    assert_eq!(
+        app.world
+            .get_resource::<ActionState<TestAction>>()
+            .unwrap()
+            .current_duration(&TestAction::Up),
+        Duration::from_millis(18)
+    );
+}
+
+/// We have a frames with two FixedUpdate schedule executions in between (F1 - FU1 - FU2 - F2)
+///
+/// A button pressed in F1 should still be marked as `just_pressed` in FU1, and should just be
+/// `pressed` in FU2
+#[test]
+fn frame_with_two_fixed_timestep() {
+    let mut app = build_app(Duration::from_millis(4), Duration::from_millis(9));
+
+    app.press_input(KeyCode::ArrowUp);
+
+    // the FixedUpdate schedule should run twice
+    // the button should be just_pressed only once
+    app.update();
+    assert_eq!(
+        app.world.get_resource::<FixedUpdateCounter>().unwrap().run,
+        2
+    );
+    assert_eq!(
+        app.world
+            .get_resource::<FixedUpdateCounter>()
+            .unwrap()
+            .just_pressed,
+        1
+    );
+    assert_eq!(
+        app.world
+            .get_resource::<UpdateCounter>()
+            .unwrap()
+            .just_pressed,
+        1
+    );
+
+    // the FixedUpdate schedule should run twice
+    app.update();
+    assert_eq!(
+        app.world.get_resource::<FixedUpdateCounter>().unwrap().run,
+        4
+    );
+    assert_eq!(
+        app.world
+            .get_resource::<FixedUpdateCounter>()
+            .unwrap()
+            .just_pressed,
```

Comment:
these checks are a bit unintuitive and repetitive, can you extract it into a helper function possibly? and make the pressed vs just_pressed more clear, it currently kinda just looks like a copy-paste oversight since its identical to the checks above. maybe have it be a reset-check of a bool, or a change tick, and assert whether its currently `just_pressed` instead of checking how many times its been `just_pressed`.

Reply:
Could you elaborate on "make the pressed vs just_pressed more clear"? I'm always only checking "just_pressed"

Reply:
I updated the tests, they should be more readable now